app/models/mixins/purging_mixin.rb

```diff
@@ -46,6 +60,57 @@ def purge_by_date(older_than, window = nil, &block)
       total
     end
 
+    def purge_count_by_remaining(remaining)
+      purge_ids_for_remaining(remaining).size
+    end
+
+    # @param [Integer] remaining number of records per resource to keep (remains in table)
+    # @param [Integer] window number of records to delete in a batch
+    # @return [Integer] number of records deleted
+    def purge_by_remaining(remaining, window = nil, &block)
+      _log.info("Purging #{table_name.humanize} older than last #{remaining} results...")
+      total = purge_in_batches(purge_ids_for_remaining(remaining), window || purge_window_size, &block)
+      _log.info("Purging #{table_name.humanize} older than last #{remaining} results...Complete - Deleted #{total} records")
+      total
+    end
+
+    # Private:  The ids to purge if we want to keep a fixed number of records
+    # for each resource. Newer records (with a higher id/ lower rank) are kept.
+    #
+    # example:
+    #
+    # So given the following DriftStates:
+    #
+    # {id: 1, resource_type: 'foo', resource_id: 1}  # Rank = 2
+    # {id: 2, resource_type: 'foo', resource_id: 1}  # Rank = 1
+    # {id: 3, resource_type: 'foo', resource_id: 2}  # Rank = 2
+    # {id: 4, resource_type: 'foo', resource_id: 2}  # Rank = 1
+    # {id: 5, resource_type: 'bar', resource_id: 6}  # Rank = 1
+    # {id: 6, resource_type: 'baz', resource_id: 4}  # Rank = 1
+    #
+    # For a "remaining" value of `1`, id 1 and 3 would then be up for deletion.
+    #
+    # @param [Integer] remaining number of records per resource to keep (remains in table)
+    # @return [Scope] Records to be deleted
+    def purge_ids_for_remaining(remaining)
+      # NOTE: `as(table_name)` fixes a bug with pluck(:id)
+      from(purge_ids_ranked_by_age.as(table_name)).where("rank > ?", remaining)
+    end
+
+    # Private: Rank records by age for each resource
+    #
+    # Assigns a "RANK" value to each record, categorized by their foreign key(s).
+    #
+    # The higher the id, the lower the rank value, the higher the priority.
+    #
+    # The values to purge are the records with a rank greater than the number
+    # of records to keep.
+    #
+    # @return [Scope] Scope of the ids prioritized by age.
+    def purge_ids_ranked_by_age
+      select(%{"id", RANK() OVER(PARTITION BY #{purge_remaining_foreign_key} ORDER BY "id" DESC) as "rank"})
```

Comment:
I agree with @Fryguy on this one, and only want to further suggest using `sanitize_sql` to do most of that heavy lifting with the `Array.warp`:

```ruby
sql = <<--EOSQL
  "id", RANK() OVER(
    PARTITION BY :foreign_key
    ORDER BY "id" DESC
  ) AS "rank"
EOSQL
select(sanitize_sql [sql, :foreign_key => purge_remaining_foreign_key])
```

This will single quote each of the values in the array, but that should be fine, and I believe that is still valid SQL.